x/wasm/internal/keeper/ioutil_test.go

```diff
@@ -0,0 +1,63 @@
+package keeper
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestUncompress(t *testing.T) {
+	wasmRaw, err := ioutil.ReadFile("./testdata/contract.wasm")
+	require.NoError(t, err)
+
+	wasmGzipped, err := ioutil.ReadFile("./testdata/contract.wasm.gzip")
+	require.NoError(t, err)
+
+	specs := map[string]struct {
+		src       []byte
+		expError  error
+		expResult []byte
+	}{
+		"handle wasm uncompressed": {
+			src:       wasmRaw,
+			expResult: wasmRaw,
+		},
+		"handle wasm compressed": {
+			src:       wasmGzipped,
+			expResult: wasmRaw,
+		},
+		"handle nil slice": {
+			src:       nil,
+			expResult: nil,
+		},
+		"handle short unidentified": {
+			src:       []byte{0x1, 0x2},
+			expResult: []byte{0x1, 0x2},
+		},
+		"handle big slice": {
+			src:       []byte(strings.Repeat("a", maxSize+1)),
+			expResult: []byte(strings.Repeat("a", maxSize+1)),
+		},
+		"handle gzip identifier only": {
+			src:      gzipIdent,
+			expError: io.ErrUnexpectedEOF,
+		},
+		"handle broken gzip": {
```

Comment:
Please add a test or two for zip bombs (after reading your comment, I'm thinking a lot about them)

I found the following famous zip bomb (42.zip): https://www.unforgettable.dk/ but it uses many files (recursively), so this might be rejected anyway (multistream=false).

Here is an article on non-recursive zip bombs, along with some links. One claims  42 kB → 5.5 GB

https://www.bamsoftware.com/hacks/zipbomb/ ([original paper](https://www.usenix.org/system/files/woot19-paper_fifield_0.pdf))

I'm not sure if these work on gzip, but it would be good to find a proper copy to validate the defense.